### ISSUE TITLE
Like `manyUntil`, but gracefully accepts EOF

### ISSUE THREAD

**Issue by mlms13** [reazen/relude-parse#18]
In relude-csv, I've found myself wanting to parse either until a terminator is found, or until all input is consumed. With `manyUntil`'s current behavior, the terminator is required or parsing will fail. I don't think this is wrong, but it's a little verbose to work around:

```reason
manyUntil(str(";") |> map(ignore) <|> eof);
```

Maybe something like `manyUntilOpt` since the terminator isn't required to be present for the parser to succeed?

**Comment by andywhite37** [comment]
Yeah, I think that makes sense

**Comment by andywhite37** [comment]
Another random thing to add might be an `orEOF` function which takes a parser, which returns any type of value, and runs the parser, and throws away the result, and if it fails, try the eof parser.

That way, you could do:

```
manyUntil(str(";") |> orEOF)
```

Not super elegant, but might be handy.

As a side note, a couple shorthand versions of the `map(ignore)` example would be:

```re
manyUntil(str(";") |> void <|> eof) // using the void Functor extension
manyUntil(str(";") $> () <|> eof) // using the voidRight Functor extension
```

But neither of these are much shorter nor more readable than the `map(ignore)`

**Comment by mlms13** [comment]
I forgot about `void`... I do like that `|> void` option. `orEOF` could also be useful, and definitely easier to read than what I currently have. Maybe a specialized `manyUntil` wouldn't even be needed at that point. If you _do_ want to get into the business of specializing `manyUntil`, I have lots of ideas...

- `manyUntilStr` == `manyUntil(str(_))`
- `manyUntilAnyOf` == `manyUntil(anyOf(_))`
- `manyUntilAnyOfStr` == `manyUntil(anyOfStr(_))`
- `anyCharUntil` == `manyUntil(_, anyChar)`

**Comment by andywhite37** [comment]
I added `orEOF` and `orEOL` (and some various line-ending parsers) in `v0.6.0`.

I also added the `manyUntilPeek` functions that work like `manyUntil` but don't consume the terminator.

I'd like to avoid adding lots of specialized functions, as I already have some specialized functions that I regret, and I don't want to make the library grow so huge, but the four functions you mentioned above seem useful enough to maybe warrant being added.  However, since they are so easy to compose as needed, maybe I'll not add those at this time.

I'll close this issue for now, and we can open a new issue if you run into more hurdles in `relude-csv`.
